### DISCUSSION TITLE
[Feature] Enhanced Memories (Trips)

### DISCUSSION THREAD

**Discussion by arnolicious** [immich-app/immich#7437] [Feature Request]
### The feature

In _The App which shall not be named_ there are not only memories for pictures taken on the specific day in previous years, but also others. Some seem quite arbitrary ("Remember this day", "Revisit the momet", "Spotlight on a day") that I don't really find valuable, since they do seem to pick out pretty random pictures, not linked to the current Date.

However one type that I do enjoy are the "Trip to [location]" Memories. These recognize trips, by combining the dates and geolocation of pictures, and show this Memory, seemingly during the dates of that trip.

This should be possible with the current data, and I figure the logic to recognize these trips shouldn't be too complex to implement.

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by DawidPietrykowski** [comment]
Good idea! I would also like to add that I personally find "Remember this day" or "<Random Person> memories" quite nice and these also shouldn't be too hard to implement.
I might try to implement it

**Reply by arnolicious** [reply]
Any idea how those pictures get chosen? Especially the "Remember this day" ones, I haven't been able to figure out any logic for their selction...

**Reply by DawidPietrykowski** [reply]
Hard to tell, I'm also unsure how do they select them, but the ones I got were often quite good (not like random photos taken around the flat but events, trips, etc) so I assume they have some criteria. 
The days I got often had many photos so I assume that it filters out days which have less than maybe 20 photos.

I would propose a system which filters out obvious uninteresting days (low photo count) and ranks other ones based on a few criterias:
- from how many locations do the picture come from?
- at what time were the photos taken? (more spread out throughout the day the better)
- how many people are in these photos? (the more the better) 
- search based on embedding of positive keywords like: love, fun, joy, family, friends, etc. (tested it right now and seems to work quite well)

then we could pick a random one from those ranked as "good"

What do you think?

**Reply by arnolicious** [reply]
That does sound like a very solid base to start on 👍

**Comment by dukobpa3** [comment]
I'll add my own 5 cents

Would be great to generate "Events" albums. Each app have this feature with more of less valuable implementation.
- photoprism makes events more abstract like: "towers", "spain" etc. 
- apple photos also, but this one the last in my AI rating because it is so dummy :)
- google photos doing it the best for now, but even their implementation is not ideal.

Last one I liked is from `librephotos`. This app is so unstable and buggy, but I really like their autogenerated albums based on time + geo. Looks like there is not so complicated logic, would be great to start from that point and then maybe improve step by step. It is opensource, so we can even take some their ideas somehow.

As for me this feature is most valuable in case of photo organising. That's why I can't move from google for the many years since picaseweb app. I'm organising my photos like stream + albums. And very important to have some automated suggestions of new albums, to avoid some manual monkeyjob.

This is explanation from `librephotos` library management page.
![Screenshot 2024-04-06 at 17 38 09](https://github.com/immich-app/immich/assets/1858880/04777f6d-7c6e-45e1-b7cb-572c9c36eceb)

This is result. Not ideal, but good enough. For example sometimes it makes one bigger trip as couple smaller etc. But anyway it is good start to moderate it later and keep some of them with some modifications like add my own name manually, merge couple generated albums to one valuable, keep some of them, delete others etc.
![Screenshot 2024-04-06 at 17 51 20](https://github.com/immich-app/immich/assets/1858880/65904b32-821a-4ec0-baea-062f611f8d5b)


And also there can be some memories in the top bar based on similar logic.

**Reply by dukobpa3** [reply]
And also librephotos have separate menus for user albums and AI albums. So in their case good option to generate many AI albums and then save some of them to user-defined. Works like kind of AI suggestions.
![Screenshot 2024-04-06 at 18 05 01](https://github.com/immich-app/immich/assets/1858880/cc9d7f60-b593-412e-a23e-296b13571e9b)

**Reply by dukobpa3** [reply]
And for example similar result from photoprism
As for me it is much less useful because there is not "Trips", just merged location and date. There is Hungary 2016, Germany 2016, Slovakia 2016, Slovenia 2016 was one trip. And Spain 2016 another one.
Librephotos recognised it well
![Screenshot 2024-04-06 at 17 07 54](https://github.com/immich-app/immich/assets/1858880/fb28e377-66a3-4702-a21e-7d2a41d5c122)

**Reply by dukobpa3** [reply]
In case of google – it understand your main location, and something "not regular". So even if you drove somewhere about 50-100km for about couple hours from your regular place – it can understand there was a "trip".

**Comment by petersuba** [comment]
Also see #2836 which is wider than this but does not go to much detail in this specific type of a Memory. 
In terms of selecting the "right" photos, ML and sentiment analysis should help - I assume the Smart Search would have generated data that could be used for sentiment analysis and helping with the selection of good photos (i.e. photos that are sharp, have smiling people or nice nature photos/landscapes, good lighting and colours etc).
Geolocation and especially grouping of photos to nearby locations that are not typical in the timeline should be a good way to find "trips".

**Comment by qrkourier** [comment]
Similarly, and perhaps more simply, I'd like to interact with memories. For example, when confronted with a memory, it's impossible to interact with the Immich entity from which the memory was created. I only wish to view the item, and to explore it in the context of the timeline where it appears, but I am trapped inside the "memories" part of the UI.

**Reply by arnolicious** [reply]
Well, it should be possible to jump to the date of the memory, and from there you can interact.

I implemented that in #7323

**Reply by qrkourier** [reply]
Ah, neat. I see that now in the mobile UI. I was looking for something like that in the web UI.

**Reply by anandv96** [reply]
As of 1.113 this is on the Web UI as well [https://github.com/immich-app/immich/pull/12117](https://github.com/immich-app/immich/pull/12117)

**Comment by CrushedAsian255** [comment]
I think a better example are the Memories generated by Apple Photos, as their Memories seem to be more coherent than *App-which-shall-not-be-named*.

**Reply by dukobpa3** [reply]
Sorry but IMO apple photos dummiest from all available galleries opensource and proprietary :)

I'm using Google and Apple at the same time with the same library for a couple years and Apple does literally nothing special(or just nothing).

**Reply by CrushedAsian255** [reply]
In most features, yes Apple photos can be a pain. However I do genuinely think their Memories and Trips classification system is pretty well thought out.

**Comment by DrD4ffy** [comment]
I like this idea. From a scientific standpoint it sounds like a 'Cluster analysis' problem. Find distinct clusters (bunch of photos) in space and time. Simple (and obvious example): Last year in October I was in Paris for seven days. All photos of this trip will be identified as a cluster because they are close together in space (geolocation) and time. A photo of my cat at home, on the day of departure would probably not be included because its in the same time frame but at a significant different location.
[Wikipedia - Cluster analysis](https://en.wikipedia.org/wiki/Cluster_analysis#Algorithms)

**Reply by CrushedAsian255** [reply]
Immich already uses clustering for the ML features, so this could probably work and be implemented without a complete overhaul of the system. Maybe a new job for this clustering

**Comment by zoro273** [comment]
Related Feature Request 
https://github.com/immich-app/immich/discussions/2836
